dot/sync/state_sync.go

```diff
@@ -114,90 +159,114 @@ func (s *StateSyncStrategy) Process(results []*SyncTaskResult) (
-			if err != nil {
-				switch err {
-				case ErrEmptyStateEntries:
-					logger.Infof("Bad state response")
-					peersToBlock = append(peersToBlock, result.who)
-					repChanges = append(repChanges, Change{
-						who: result.who,
-						rep: peerset.ReputationChange{
-							Value:  peerset.BadStateValue,
-							Reason: peerset.BadStateReason,
-						},
-					})
-					continue
+	switch s.phase {
+	case DownloadState:
+		for _, result := range results {
+			switch response := result.response.(type) {
+			case *messages.StateResponse:
+				logger.Debugf("Retrieving state data from %s with %d keys",
+					result.who, len(response.Entries))
+
+				completed, err := s.stateRequestProvider.ProcessResponse(response)
+				if err != nil {
+					switch err {
+					case ErrEmptyStateEntries:
+						logger.Infof("Bad state response")
+						peersToBlock = append(peersToBlock, result.who)
+						repChanges = append(repChanges, Change{
+							who: result.who,
+							rep: peerset.ReputationChange{
+								Value:  peerset.BadStateValue,
+								Reason: peerset.BadStateReason,
+							},
+						})
+						continue
+					}
 				}
-			}
 
-			if s.completed {
-				return true, repChanges, peersToBlock, s.importState()
-			}
+				if completed {
+					s.phase = DownloadFirstBlock
+				}
 
-			return false, repChanges, peersToBlock, nil
-
-		default:
-			logger.Warnf("unexpected response type %T for state request, banning peer: %s", result.who, response)
-			repChanges = append(repChanges, Change{
-				who: result.who,
-				rep: peerset.ReputationChange{
-					Value:  peerset.UnexpectedResponseValue,
-					Reason: peerset.UnexpectedResponseReason,
-				}})
-			peersToBlock = append(peersToBlock, result.who)
-			continue
+				return false, repChanges, peersToBlock, nil
+			default:
+				logger.Warnf("unexpected response type %T for state request, banning peer: %s", response, result.who)
+				repChanges = append(repChanges, Change{
+					who: result.who,
+					rep: peerset.ReputationChange{
+						Value:  peerset.UnexpectedResponseValue,
+						Reason: peerset.UnexpectedResponseReason,
+					}})
+				peersToBlock = append(peersToBlock, result.who)
+				continue
+			}
 		}
-	}
-
-	return s.IsSynced(), repChanges, peersToBlock, nil
-}
+	case DownloadFirstBlock:
+		var validRes []RequestResponseData
 
-// importState imports the retreived state into our state storage
-func (s *StateSyncStrategy) importState() error {
-	// Store state in our state storage
-	trieState, err := s.stateRequestProvider.BuildTrie()
-	if err != nil {
-		return err
-	}
+		// Reuse same validator than in fullsync
+		repChanges, peersToBlock, validRes = validateResults(results, s.badBlocks)
 
-	if trieState.MustHash() != s.targetBlock.StateRoot {
-		return fmt.Errorf("state root mismatch: got %s expected %s", trieState.MustHash(), s.targetBlock.StateRoot)
+		if len(validRes) > 0 && validRes[0].responseData != nil && len(validRes[0].responseData) > 0 {
```

Comment:
we can do just `len(validRes[0].responseData)` since `len(nil) == 0`
```suggestion
		if len(validRes) > 0 &&  len(validRes[0].responseData) > 0 {
```